crates/bevy_ecs/src/reflect/bundle.rs

```diff
@@ -274,3 +168,57 @@ impl<C: Bundle + Reflect + FromWorld> FromType<C> for ReflectBundle {
         })
     }
 }
+
+fn insert_field<C: 'static>(entity: &mut EntityMut, field: &dyn Reflect, registry: &TypeRegistry) {
+    if let Some(reflect_component) = registry.get_type_data::<ReflectComponent>(field.type_id()) {
+        reflect_component.apply(entity, field);
+    } else if let Some(reflect_bundle) = registry.get_type_data::<ReflectBundle>(field.type_id()) {
+        reflect_bundle.apply(entity, field, registry);
+    } else {
+        entity.world_scope(|world| {
+            if let Some(id) = world.bundles().get_id(TypeId::of::<C>()) {
+                let info = world.bundles().get(id).unwrap();
+                if info.components().is_empty() {
+                    panic!(
+                        "no `ReflectComponent` registration found for `{}`",
+                        field.type_name()
+                    );
+                }
+            };
+        });
+
+        panic!(
+            "no `ReflectBundle` registration found for `{}`",
+            field.type_name()
+        )
+    }
+}
+
+fn apply_or_insert_field<C: 'static>(
```

Comment:
```suggestion
fn apply_or_insert_field<B: 'static>(
```
Nit, non-blocking: I would use `B` instead of `C` everywhere in this file.